presto-main/src/main/java/com/facebook/presto/sql/gen/JoinFilterFunctionCompiler.java

```diff
@@ -0,0 +1,190 @@
+    {
+        CachedInstanceBinder cachedInstanceBinder = new CachedInstanceBinder(classDefinition, callSiteBinder);
+        generateFilterMethod(classDefinition, callSiteBinder, cachedInstanceBinder, filter, leftBlocksSize);
+        generateConstructor(classDefinition, cachedInstanceBinder);
+    }
+
+    private void generateFilterMethod(ClassDefinition classDefinition, CallSiteBinder callSiteBinder, CachedInstanceBinder cachedInstanceBinder, RowExpression filter, int leftBlocksSize)
+    {
+        // todo handle TRY expression
+        Map<CallExpression, MethodDefinition> tryMethodMap = ImmutableMap.of();
+
+        //int leftPosition, Block[] leftBlocks, int rightPosition, Block[] rightBlocks
+        Parameter leftPosition = arg("leftPosition", int.class);
+        Parameter leftBlocks = arg("leftBlocks", Block[].class);
+        Parameter rightPosition = arg("rightPosition", int.class);
+        Parameter rightBlocks = arg("rightBlocks", Block[].class);
+
+        MethodDefinition method = classDefinition.declareMethod(
+                a(PUBLIC),
+                "filter",
+                type(boolean.class),
+                ImmutableList.<Parameter>builder()
+                        .add(leftPosition)
+                        .add(leftBlocks)
+                        .add(rightPosition)
+                        .add(rightBlocks)
+                        .build());
+
+        method.comment("filter: %s", filter.toString());
+        BytecodeBlock body = method.getBody();
+
+        Scope scope = method.getScope();
+        Variable wasNullVariable = scope.declareVariable("wasNull", body, constantFalse());
+
+        BytecodeExpressionVisitor visitor = new BytecodeExpressionVisitor(
+                callSiteBinder,
+                cachedInstanceBinder,
+                fieldReferenceCompiler(callSiteBinder, leftPosition, leftBlocks, rightPosition, rightBlocks, leftBlocksSize, wasNullVariable),
+                metadata.getFunctionRegistry(),
+                tryMethodMap);
+
+        BytecodeNode visitorBody = filter.accept(visitor, scope);
+
+        Variable result = scope.declareVariable(boolean.class, "result");
+        body.append(visitorBody)
+                .putVariable(result)
+                .append(new IfStatement()
+                        .condition(wasNullVariable)
+                        .ifTrue(constantFalse().ret())
+                        .ifFalse(result.ret()));
+    }
+
+    private static RowExpressionVisitor<Scope, BytecodeNode> fieldReferenceCompiler(
+            final CallSiteBinder callSiteBinder,
+            final Variable leftPosition,
+            final Variable leftBlocks,
+            final Variable rightPosition,
+            final Variable rightBlocks,
+            final int leftBlocksSize,
+            final Variable wasNullVariable)
+    {
+        return new RowExpressionVisitor<Scope, BytecodeNode>()
```

Comment:
I'd like to see this and PageProcessor

The new class would be
```java
import com.facebook.presto.bytecode.BytecodeNode;
import com.facebook.presto.bytecode.Scope;
import com.facebook.presto.bytecode.Variable;
import com.facebook.presto.bytecode.control.IfStatement;
import com.facebook.presto.bytecode.expression.BytecodeExpression;
import com.facebook.presto.spi.type.Type;
import com.facebook.presto.sql.relational.CallExpression;
import com.facebook.presto.sql.relational.ConstantExpression;
import com.facebook.presto.sql.relational.InputReferenceExpression;
import com.facebook.presto.sql.relational.RowExpressionVisitor;
import com.google.common.primitives.Primitives;
import io.airlift.slice.Slice;

import java.util.function.BiFunction;

import static com.facebook.presto.sql.gen.SqlTypeBytecodeExpression.constantType;
import static java.util.Objects.requireNonNull;

class InputReferenceCompiler
        implements RowExpressionVisitor<Scope, BytecodeNode>
{
    private final BiFunction<Scope, Integer, BytecodeExpression> blockResolver;
    private final BiFunction<Scope, Integer, BytecodeExpression> positionResolver;
    private final Variable wasNullVariable;
    private final CallSiteBinder callSiteBinder;

    public InputReferenceCompiler(
            BiFunction<Scope, Integer, BytecodeExpression> blockResolver,
            BiFunction<Scope, Integer, BytecodeExpression> positionResolver,
            Variable wasNullVariable,
            CallSiteBinder callSiteBinder)
    {
        this.blockResolver = requireNonNull(blockResolver, "blockResolver is null");
        this.positionResolver = requireNonNull(positionResolver, "positionResolver is null");
        this.wasNullVariable = requireNonNull(wasNullVariable, "wasNullVariable is null");
        this.callSiteBinder = requireNonNull(callSiteBinder, "callSiteBinder is null");
    }

    @Override
    public BytecodeNode visitInputReference(InputReferenceExpression node, Scope scope)
    {
        int field = node.getField();
        Type type = node.getType();

        BytecodeExpression block = blockResolver.apply(scope, field);
        BytecodeExpression position = positionResolver.apply(scope, field);

        Class<?> javaType = type.getJavaType();
        if (!javaType.isPrimitive() && javaType != Slice.class) {
            javaType = Object.class;
        }

        IfStatement ifStatement = new IfStatement();
        ifStatement.condition(block.invoke("isNull", boolean.class, position));

        ifStatement.ifTrue()
                .putVariable(wasNullVariable, true)
                .pushJavaDefault(javaType);

        String methodName = "get" + Primitives.wrap(javaType).getSimpleName();
        ifStatement.ifFalse(constantType(callSiteBinder, type).invoke(methodName, javaType, block, position));

        return ifStatement;
    }

    @Override
    public BytecodeNode visitCall(CallExpression call, Scope scope)
    {
        throw new UnsupportedOperationException("not yet implemented");
    }

    @Override
    public BytecodeNode visitConstant(ConstantExpression literal, Scope scope)
    {
        throw new UnsupportedOperationException("not yet implemented");
    }
}
``` 

And the usage in PageProcessor would be:
```java
    private static RowExpressionVisitor<Scope, BytecodeNode> fieldReferenceCompiler(final CallSiteBinder callSiteBinder, final Variable positionVariable, final Variable wasNullVariable)
    {
        return new InputReferenceCompiler(
                (scope, field) -> scope.getVariable("block_" + field),
                (scope, field) -> positionVariable,
                wasNullVariable,
                callSiteBinder);
    }
```

And the usage in this class would be this variable declaration (it is simpler than a method for a single call):
```java
        RowExpressionVisitor<Scope, BytecodeNode> fieldReferenceCompiler = new InputReferenceCompiler(
                (scope, field) -> field < leftBlocksSize ? leftBlocks.getElement(field) : rightBlocks.getElement(field - leftBlocksSize),
                (scope, field) -> field < leftBlocksSize ? leftPosition : rightPosition,
                wasNullVariable,
                callSiteBinder);
```